keras_nlp/models/bart/bart_backbone.py

```diff
@@ -0,0 +1,245 @@
+
+    Disclaimer: Pre-trained models are provided on an "as is" basis, without
+    warranties or conditions of any kind. The underlying model is provided by a
+    third party and subject to a separate license, available
+    [here](https://github.com/facebookresearch/fairseq/tree/main/examples/bart).
+
+    Args:
+        vocabulary_size: int. The size of the token vocabulary.
+        num_layers: int. The number of transformer encoder layers and
+            transformer decoder layers.
+        num_heads: int. The number of attention heads for each transformer.
+            The hidden size must be divisible by the number of attention heads.
+        hidden_dim: int. The size of the transformer encoding and pooler layers.
+        intermediate_dim: int. The output dimension of the first Dense layer in
+            a two-layer feedforward network for each transformer.
+        dropout: float. Dropout probability for the Transformer encoder.
+        max_sequence_length: int. The maximum sequence length that this encoder
+            can consume. If None, `max_sequence_length` uses the value from
+            sequence length. This determines the variable shape for positional
+            embeddings.
+
+    Examples:
+    ```python
+    input_data = {
+        "encoder_token_ids": tf.ones(shape=(1, 12), dtype=tf.int64),
+        "encoder_padding_mask": tf.constant(
+            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)
+        ),
+        "decoder_token_ids": tf.ones(shape=(1, 12), dtype=tf.int64),
+        "decoder_padding_mask": tf.constant(
+            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)
+        ),
+    }
+
+    # Randomly initialized BART encoder-decoder model with a custom config
+    model = keras_nlp.models.BartBackbone(
+        vocabulary_size=50265,
+        num_layers=6,
+        num_heads=12,
+        hidden_dim=768,
+        intermediate_dim=3072,
+        max_sequence_length=12,
+    )
+    output = model(input_data)
+    ```
+    """
+
+    def __init__(
+        self,
+        vocabulary_size,
+        num_layers,
+        num_heads,
+        hidden_dim,
+        intermediate_dim,
+        dropout=0.1,
+        max_sequence_length=1024,
+        **kwargs,
+    ):
+
+        # Encoder inputs
+        encoder_token_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="encoder_token_ids"
+        )
+        encoder_padding_mask = keras.Input(
+            shape=(None,), dtype="int32", name="encoder_padding_mask"
+        )
+
+        # Decoder inputs.
+        decoder_token_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="decoder_token_ids"
+        )
+        decoder_padding_mask = keras.Input(
+            shape=(None,), dtype="int32", name="decoder_padding_mask"
+        )
+
+        # Token embedding layer. This layer is shared by encoder and decoder.
+        token_embedding_layer = keras.layers.Embedding(
```

Comment:
When BART is used as a language model, are the embedding weights shared for the output projection? Or is there a separate set of parameters used?

Reply:
@mattdangerw - they do define separate Embedding layers in both the encoder and decoder: https://github.com/huggingface/transformers/blob/44caf4f6f47120a4aaca561c9cc8041455bef705/src/transformers/models/bart/modeling_bart.py#L723-L726 and https://github.com/huggingface/transformers/blob/44caf4f6f47120a4aaca561c9cc8041455bef705/src/transformers/models/bart/modeling_bart.py#L896-L899.

But they copy over the weights for these layers from this layer: https://github.com/huggingface/transformers/blob/main/src/transformers/models/bart/modeling_bart.py#L1169. 

Moreover, when I printed the trainable layers, these two embedding layers did not show up: https://colab.research.google.com/drive/15871_w5LqU9-wzykFbb8dAne6g4zvyat?usp=sharing. So, I think the same set of parameters are being used?

Reply:
Great! Great, this is the easy case for us, where the backbone is self sufficient for language modeling tasks. We have prior art for this with BERT pretraining and Chen's upcoming GPT2 language model.